backend/accounts/transaction.go

```diff
@@ -133,5 +142,122 @@ type OrderedTransactions []*TransactionData
 // The input list is modified in place and must not be used anymore after calling this function.
 func NewOrderedTransactions(txs []*TransactionData) OrderedTransactions {
 	sort.Sort(sort.Reverse(byHeight(txs)))
+
+	balance := big.NewInt(0)
+	for i := len(txs) - 1; i >= 0; i-- {
+		tx := txs[i]
+		switch tx.Type {
+		case TxTypeReceive:
+			balance.Add(balance, tx.Amount.BigInt())
+		case TxTypeSend:
+			balance.Sub(balance, tx.Amount.BigInt())
+			// Subtract fee as well.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		case TxTypeSendSelf:
+			// Subtract only fee.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		}
+		tx.Balance = coin.NewAmount(balance)
+	}
 	return txs
 }
+
+// TimeseriesEntry contains the balance of the account at the given time.
+type TimeseriesEntry struct {
+	Time  time.Time
+	Value coin.Amount
+}
+
+// MarshalJSON serializes the entry as JSON.
+func (c *TimeseriesEntry) MarshalJSON() ([]byte, error) {
+	int64, err := c.Value.Int64()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(map[string]interface{}{
+		"time":  c.Time.Unix(),
+		"value": int64,
+	})
+}
+
+// findLastIndexBeforeTime returns the index of the last transaction with a time before `target`,
+// starting at `start`.
+// Unconfirmed txs are ignored.
+// If there is no such transaction before `target`, nil is returned.
+// Note: the transactions are ordered from newest to oldest, so the returned index will be smaller
+// or equal to `start` (advancing backwards).
+func (txs OrderedTransactions) findLastIndexBeforeTime(
+	start int, target time.Time) (*int, error) {
+	currentIndex := start
+	for {
+		tx := txs[currentIndex]
+
+		if tx.Height == 0 {
+			// We reached the unconfirmed txs. We ignore those. The current index is the last tx
+			// before the target time.
+			return &currentIndex, nil
+		}
+		txTime := tx.Timestamp
+		if txTime == nil {
+			return nil, errp.WithStack(errors.ErrNotAvailable)
+		}
+		if target.Before(*txTime) {
+			if currentIndex == start {
+				// No tx before target, return nil.
+				return nil, nil
+			}
+			// We reached the first tx beyond target. Return the previous index, which is the last
+			// tx before target.
+			currentIndex++
+			return &currentIndex, nil
+		}
+		if currentIndex == 0 {
+			return &currentIndex, nil
```

Comment:
what if it's still the case that `txTime < target`? then according to the description, it should `return nil, nil`.

Reply:
if txtime < target, this is the last tx before target.

Say target is one year from now, and your latest tx (index = 0) is today. Then `0` is the last tx before target.

Reply:
> if txtime < target, this is the last tx before target.

but there's nothing else "before" the target. it's the last element.

consider this: t1, t2, t3, and target is `t1 - 1`. if `currentIndex == 0`, so we're at t1, the currentIndex still doesn't point to an item with timestamp < target. in other words: no such transaction.

or am I missing something?

Reply:
Fixed edge case where target==txTime, changing from `target.Before()` to `!target.After()`.

Still thinking about the rest.